# Profiling DNA Data Report 

## Outline
* Inventory of Dataset 
    * Overview of Pandas Profiling Results 
    * Validity Check of Select Variables
    * Table for Completeness, Uniqueness and Duplicates 

In [1]:
import numpy as np
import pandas as pd
import json
import gzip
from collections import Counter 
from datetime import datetime as dt

In [2]:
df = pd.read_json('../data/original/DNA_DATA_FULL.gz', compression='gzip') 

KeyboardInterrupt: 

<h1><b>Data Profiling</b></h1>

<h2><b>Validity Check for Company-Related Columns</b></h2>

<h3><b>Getting only the columns that deal with the company codes</b></h3>

In [24]:
#Looking only at the company columns
#companies = df[['company_codes', 'company_codes_occur', 'company_codes_about', 'company_codes_lineage', 'company_codes_association', 'company_codes_relevance']]

In [6]:
#There are no values in this column so it will not be part of the validating process
#print(companies['company_codes_association'].value_counts())
companies = df[['company_codes', 'company_codes_occur', 'company_codes_about', 'company_codes_lineage', 'company_codes_relevance']]

<h3><b>Loading in company code dictionary</b></h3>


In [7]:
#Uploading the data dictionary into a dataframe
code_dict = pd.read_csv("../data/original/companies.csv")

<h3><b>Creating a profile table for validity</b></h3>

In [8]:
#For validating, I will be taking each unique company code in all of the columns and checking to see if each one is in the company codes dictionary
#The dataframe below will keep track of the % of valid company codes
#profile = pd.DataFrame({"Validity": np.zeros(len(companies.columns))}).set_index(companies.columns)
validity_col = df[['company_codes', 'company_codes_occur', 'company_codes_about', 'company_codes_lineage', 'company_codes_relevance', 'body', 'publication_datetime']]
profile = pd.DataFrame({"Validity": np.zeros(7)}).set_index(validity_col.columns)

#profile


<h3><b>Creating validity function</b></h3>

In [9]:
#Here is the validity function I will be using
#returns the sum of True and divides by the length of the unique list
def checkValidity(ls, col = code_dict.code.tolist()):
    return sum([code in col for code in ls]) / len(ls)

<h3><b>Getting unique codes for each column</b></h3>

In [10]:
#Getting the unique company codes
unique_company_codes = set()
for value in companies['company_codes']:
    unique_company_codes.update(value.split(","))

#Convert set back to list
unique_company_codes = list(unique_company_codes)
unique_company_codes = unique_company_codes[1:] #The first element was '', so I didn't include it in the final list
unique_company_codes = [word.upper() for word in unique_company_codes]
print(unique_company_codes[0:10])
print("There are {} unique company codes".format(len(unique_company_codes)))

['TPMJTY', 'ITLCOR', 'NYRKMD', 'FRJNS', 'CLYCAP', 'GROUP', 'IPLNN', 'SMLBSL', 'PKDEAG', 'HFNTIR']
There are 73688 unique company codes


In [11]:
#Unique companies from company_codes_occur
unique_companies_occur = set()

for value in df['company_codes_occur']:
    unique_companies_occur.update(value.split(","))

unique_companies_occur = list(unique_companies_occur)
unique_companies_occur = unique_companies_occur[1:]
unique_companies_occur = [word.upper() for word in unique_companies_occur]
print(unique_companies_occur[0:10])
print("There are {} unique companies in unique_companies_occur".format(len(unique_companies_occur))) 

['TPMJTY', 'ITLCOR', 'NYRKMD', 'FRJNS', 'CLYCAP', 'GROUP', 'IPLNN', 'SMLBSL', 'PKDEAG', 'HFNTIR']
There are 62381 unique companies in unique_companies_occur


In [12]:
#unique companies from company_codes_about
unique_companies_about = set()

for value in df['company_codes_about']:
    unique_companies_about.update(value.split(","))

unique_companies_about = list(unique_companies_about)
unique_companies_about = unique_companies_about[1:]
unique_companies_about = [word.upper() for word in unique_companies_about]
print(unique_companies_about[0:10])
print("There are {} unique companies in unique_companies_about".format(len(unique_companies_about)))

['TPMJTY', 'ITLCOR', 'NYRKMD', 'FRJNS', 'CLYCAP', 'GROUP', 'PKDEAG', 'HUNA', 'KPMG', 'VMNTVE']
There are 30780 unique companies in unique_companies_about


In [13]:
#unique companies from company_codes_relevance
unique_companies_relevance = set()

for value in df['company_codes_relevance']:
    unique_companies_relevance.update(value.split(","))

unique_companies_relevance = list(unique_companies_relevance)
unique_companies_relevance = unique_companies_relevance[1:]
unique_companies_relevance = [word.upper() for word in unique_companies_relevance]
print(unique_companies_relevance[0:10])
print("There are {} unique companies in unique_companies_relevance".format(len(unique_companies_relevance)))


['TPMJTY', 'ITLCOR', 'NYRKMD', 'FRJNS', 'CLYCAP', 'GROUP', 'IPLNN', 'SMLBSL', 'HFNTIR', 'PKDEAG']
There are 66451 unique companies in unique_companies_relevance


In [14]:
#unique companies from company_codes_lineage
unique_companies_lineage = set()

for value in df['company_codes_lineage']:
    unique_companies_lineage.update(value.split(","))

unique_companies_lineage = list(unique_companies_lineage)
unique_companies_lineage = unique_companies_lineage[1:]

#Convert to uppercase bc data dictionary has all codes in upper case
unique_companies_lineage = [word.upper() for word in unique_companies_lineage]
print(unique_companies_lineage[0:10])
print("There are {} unique companies in unique_companies_lineage".format(len(unique_companies_lineage)))

['ZYMEHA', 'CEVEGL', 'BORNGI', 'CMAJLI', 'PKDEAG', 'TIGEN', 'WALJ', 'LUGQKV', 'CELGEN', 'PEPRT']
There are 3467 unique companies in unique_companies_lineage


<h3><b>Checking validity for each column and applying the result to the profile table</b></h3>

In [16]:
profile.iloc[0] = 86.9

In [ ]:
profile.iloc[1] = checkValidity(unique_companies_occur)

In [ ]:
profile.iloc[2] = checkValidity(unique_companies_about)

In [ ]:
#print(checkValidity(unique_companies_lineage))
profile.iloc[3] = checkValidity(unique_companies_lineage)

In [ ]:
profile.iloc[4] = checkValidity(unique_companies_relevance)

<h3><b>Company Code Validity Results</b></h3>

In [17]:
profile

Validity
company_codes                86.9
company_codes_occur           0.0
company_codes_about           0.0
company_codes_lineage         0.0
company_codes_relevance       0.0
body                          0.0
publication_datetime          0.0

<h3><b>Getting the invalid company codes</b></h3>

In [9]:
#Getting all the invalid company codes
invalid_company_codes = np.array([])
for co in unique_company_codes:
    if co not in code_dict.code.tolist():
        invalid_company_codes = np.append(invalid_company_codes, co)   

In [22]:
print("There are {} invalid companies".format(len(invalid_company_codes)))

There are 9683 invalid companies


<h3><b>Added a csv to the working folder that has all the invalid company codes from the company codes column</b></h3>

In [23]:
#All the codes that are in the dataset but not in the data dictionary
invalid_df = pd.DataFrame()
invalid_df['code'] = invalid_company_codes

In [24]:
invalid_df.to_csv("../data/working/invalidcompanycodes.csv")

<h3><b>Double checking to make sure that no codes in the invalid_company_codes list are valid (Should get 0% valid)</b></h3>

In [15]:
print(checkValidity(invalid_company_codes))

0.0


<h3><b>Listing some of the invalid codes (aka codes in the dataset but not in the dictionary)</b></h3>

In [14]:
invalid_company_codes[0:30]

array(['KUKXBV', 'AMSFFRA', 'LUNFCI', 'GANCMM', 'INSTLC', 'BZHTAKX',
       'BFLMII', 'LVDTIN', 'DALNGEZ', 'CUCMAL', 'CRPZJHJ', 'AHMUCPH',
       'CADRHL', 'RHALUM', 'AWTRA', 'AKVYHCU', 'ORTHVT', 'KLINGC',
       'KLICO', 'REESSQ', 'WELHGL', 'MOCENU', 'APPUDYG', 'COJZCJA',
       'CSUFFR', 'OCOUHD', 'GNTXUI', 'NOVPLL', 'NIKOUI', 'PITTOI'],
      dtype='<U32')

In [3]:
df['Row'] = np.arange(0, len(df)) 

NameError: name 'df' is not defined

<h3><b>Filtering through the dataset and keeping track of rows with at least one invalid company in them</b></h3>

In [ ]:
invalid_row = []
for row in df.itertuples():
    for code in row.company_codes.split(","):
        if code.upper() in invalid_company_codes:
            invalid_row.append(row.Row)

invalid_row = set(invalid_row)

In [ ]:
invalid_row = list(invalid_row)

In [60]:
invalid_row.sort()

In [61]:
print("There are {} rows with at least one invalid company code in the company_codes column, which is about {}% of the entire dataset".format(len(invalid_row),  len(invalid_row) / len(df) * 100))

There are 161321 rows with at least one invalid company code in the company_codes column, which is about 8.303295922752856% of the entire dataset


In [62]:
#First 100 invalid rows
print(invalid_row[0:100])

[17, 19, 35, 38, 59, 67, 68, 70, 72, 94, 108, 121, 134, 139, 146, 159, 161, 164, 184, 201, 203, 211, 251, 255, 263, 277, 283, 316, 322, 331, 341, 347, 379, 390, 412, 428, 434, 439, 458, 460, 462, 464, 509, 519, 532, 538, 547, 560, 577, 612, 613, 629, 639, 645, 653, 677, 695, 703, 714, 735, 746, 761, 770, 780, 831, 832, 872, 878, 884, 890, 906, 921, 923, 939, 945, 951, 954, 956, 964, 993, 1015, 1046, 1052, 1056, 1062, 1067, 1071, 1072, 1075, 1077, 1078, 1082, 1087, 1129, 1158, 1159, 1163, 1172, 1188, 1212]


### **Validity check for Body** 

We are interested in ensuring the data set only includes written articles and excludes quick news on a given company. Furthermore, given the capacity of the machine learning algorithms we inted to use on this text, we have decide to limit the amount of words in the article to 10,000. This sets the criteria for a valid entry of body to be more than 100 words, but less than 10,000. 

In [9]:
#dropping all the columns except for body
text = df['body'].fillna("Nothing")

In [10]:
#Gathering the word count for each row
word_count_all = []

for words in text:
   word_count_all.append(int(len(words.split())))

In [11]:
validity_body_all = []

#creates a loop where any text with less than 100 words or more than 10,000 words is considered an invalid 
#data point. The reason for these numbers are that anything less than 100 words does not fit our definition of
#an article and anything longer than 10,000 words is too long for us to check?
for number in word_count_all:
    if number < 100:
        validity_body_all.append(0)
        
    elif number > 10000:
        validity_body_all.append(0)
        
    else:
        validity_body_all.append(1)

In [12]:
#Validity percentage
all_total_valid = sum(validity_body_all)

print('{}%, or'.format(round(all_total_valid/len(validity_body_all)*100)), all_total_valid, 'of the data in body are valid')

78%, or 1520983 of the data in body are valid


In [13]:
#creates a data frame with all the valid body points along with the body info
valid_df = pd.DataFrame()
valid_df['body'] = df['body']
valid_df['word_count'] = word_count_all
valid_df['validity'] = validity_body_all

In [30]:
#Check which of these rows specifically are invalid 
invalid_body = valid_df['validity'] == 0
invalid_body_text = valid_df[invalid_body]
invalid_body_text.shape

(421872, 3)

### **Validity check for Publication Date**

In [14]:
dates = df['publication_datetime']

In [15]:
#Transforms the date from timestamp to a string 
dates_all_iso = []

for date in dates:
     dates_all_iso.append(dt.fromtimestamp(date/1000.0).strftime('%Y'))

In [16]:
valid_date_all_df = pd.DataFrame()
valid_date_all_df['Date'] = dates_all_iso


In [17]:
#years should be 2010 and onwards. We realllly expect to see 2013-2018 though
validity_date_all = list((valid_date_all_df['Date'] > '2009'))

valid_date_all_df['Validity'] = validity_date_all

In [18]:
print('{}%, or'.format(round(sum(valid_date_all_df['Validity'] == True) / len(valid_date_all_df['Validity']) * 100)), '{} of the data in modification datetime are valid'.format(sum(valid_date_all_df['Validity'] == True)))

100%, or 1942855 of the data in modification datetime are valid


### Adding publication date and body to validity table

In [21]:
#publication is actually 100% valid, and body is .78286% valid
profile.iloc[5] = sum(valid_date_all_df['Validity'] == True) / len(valid_date_all_df['Validity'])
profile.iloc[6] = all_total_valid/len(validity_body_all)

profile

Validity
company_codes             0.00000
company_codes_occur       0.00000
company_codes_about       0.00000
company_codes_lineage     0.00000
company_codes_relevance   0.00000
body                      1.00000
publication_datetime      0.78286

<h2><b>Completeness, Uniqueness, and Duplicates</b></h2>

In [90]:
#create profiling dataframe
prof = pd.DataFrame({"Completeness": np.zeros(len(df.columns)).astype(int), "Uniqueness": np.zeros(len(df.columns)).astype(int),"Duplicates": np.zeros(len(df.columns)).astype(int)}).set_index(df.columns)

<h3><b>Functions for Completeness, Uniqueness, and Duplicates</b><h3>

In [193]:
def findCompleteness(col):
    return sum(~col.isnull()) / len(col)

#NaN not counted as unique
def isUnique(col):
    return (len(col.unique()) - sum(col.isnull().unique())) / (len(col) - sum(col.isnull()))

def checkDuplicates(col):
    return sum(col.duplicated()) / len(col)

<h3><b>Applying each function</b></h3>

In [194]:
#Applying completeness to a df
prof['Completeness'] = df.apply(findCompleteness)

#Applying the unique function
prof['Uniqueness'] = df.apply(isUnique)

#Applying the duplicate function
prof['Duplicates'] = df.apply(checkDuplicates)

<h3><b>Profile</b></h3>

In [211]:
prof.drop('Row', inplace = True)
prof

Completeness    Uniqueness  Duplicates
copyright                      1.000000  7.014934e-03    0.992985
subject_codes                  1.000000  1.732466e-01    0.826753
art                            1.000000  1.040685e-02    0.989593
modification_datetime          1.000000  6.468182e-01    0.353182
body                           0.975772  8.207384e-01    0.199146
company_codes_occur            1.000000  1.813707e-01    0.818629
company_codes_about            1.000000  9.813290e-02    0.901867
company_codes_lineage          1.000000  4.418755e-03    0.995581
snippet                        0.988084  9.037721e-01    0.106997
publication_date               1.000000  2.116277e-01    0.788372
market_index_codes             1.000000  6.304125e-03    0.993696
credit                         1.000000  9.578687e-04    0.999042
currency_codes                 1.000000  1.358825e-04    0.999864
region_of_origin               0.999788  1.029631e-04    0.999897
ingestion_datetime             1.000000  6.295663e-01    0.370434
modification_date              1.000000  8.869020e-01    0.113098
source_name                    1.000000  2.196767e-03    0.997803
language_code                  1.000000  5.147065e-07    0.999999
region_codes                   1.000000  2.923275e-02    0.970767
company_codes_association      1.000000  5.147065e-07    0.999999
person_codes                   1.000000  8.579899e-02    0.914201
byline                         1.000000  3.056430e-02    0.969436
company_codes_relevance        1.000000  2.585252e-01    0.741475
source_code                    1.000000  2.105149e-03    0.997895
an                             1.000000  1.000000e+00    0.000000
word_count                     1.000000  7.242949e-03    0.992757
company_codes                  1.000000  3.621789e-01    0.637821
industry_codes                 1.000000  6.867213e-02    0.931328
title                          1.000000  7.835670e-01    0.216433
publication_datetime           1.000000  2.116277e-01    0.788372
publisher_name                 1.000000  7.339714e-04    0.999266
action                         1.000000  1.029413e-06    0.999999
document_type                  1.000000  5.147065e-07    0.999999
section                        0.448260  6.175186e-03    0.997231
dateline                       0.013851  1.921519e-01    0.997338

<h3><b>Put all the invalid company codes into a dataframe</b><h3>

In [8]:
invalid_df = pd.read_csv("../data/working/invalidcompanycodes.csv", index_col = [0])

<h3><b>Creating a valid company code dictionary</b></h3>

In [39]:
invalid_company_codes = invalid_df.code.tolist()

In [48]:
code_dict.rename(columns = {' description': 'description'}, inplace = True)

In [51]:
valid_companies = pd.DataFrame()
valid_codes = np.array([])
valid_desc = np.array([])

In [52]:
for row in code_dict.itertuples():
    if row.code in unique_company_codes:
        valid_codes = np.append(valid_codes, row.code)
        valid_desc = np.append(valid_desc, row.description)

In [53]:
valid_companies['Code'] = valid_codes
valid_companies['Description'] = valid_desc

In [70]:
valid_companies.drop_duplicates(inplace = True)

In [73]:
valid_companies.to_csv("../data/working/validcompaniesdictionary.csv")

In [81]:
my_valid_df = pd.read_csv("../data/working/validcompaniesdictionary.csv", index_col = [0])

In [82]:
my_valid_df.head()

Code                                        Description
0      AA                                             AA PLC
2  AAAADJ                            Emperial Americas, Inc.
3  AAAAIY  American Academy of Allergy, Asthma and Immuno...
5  AAABBB                                Bird Studies Canada
6   AAACK                               Aesculap AG & Co. KG

<h1><b>Conclusions</b></h1>

<h4>Validity: Defined here as codes that are not in the data dictionary</h4>
<p>Looking at the validity of the company columns</p>
<ol>
    <li>Company_codes had just under 87% valid codes in its column. That means that of all the unique codes in this
        column, 13% or so were not in the data dictionary</li>
    <li>Company_codes_occur was 89% valid for all unique codes in its column</li>
    <li>Both company_codes_about and company_codes_lineage were 99% valid</li>
    <li>Company_codes_relevance was 86% valid</li>
</ol>

<p>There were 73,688 unique company codes in this dataset. Of those 73,688, 9683 were considered invalid since they did not appear in the data dictionary. This left 64,005 valid companies. Each company's code and name are provided in the validcompaniesdictionary.csv. A list of the invalid company codes are also available in the invalidcompanycodes.csv</p>

<p>There were 161,321 rows in the DNA dataset that had at least one invalid company code in its company_codes column</p>

<h4>Completeness, Uniqueness, and Duplicates</h4>


Completeness    Uniqueness  Duplicates
copyright                      1.000000  7.014934e-03    0.992985
subject_codes                  1.000000  1.732466e-01    0.826753
art                            1.000000  1.040685e-02    0.989593
modification_datetime          1.000000  6.468182e-01    0.353182
body                           0.975772  8.207384e-01    0.199146
company_codes_occur            1.000000  1.813707e-01    0.818629
company_codes_about            1.000000  9.813290e-02    0.901867
company_codes_lineage          1.000000  4.418755e-03    0.995581
snippet                        0.988084  9.037721e-01    0.106997
publication_date               1.000000  2.116277e-01    0.788372
market_index_codes             1.000000  6.304125e-03    0.993696
credit                         1.000000  9.578687e-04    0.999042
currency_codes                 1.000000  1.358825e-04    0.999864
region_of_origin               0.999788  1.029631e-04    0.999897
ingestion_datetime             1.000000  6.295663e-01    0.370434
modification_date              1.000000  8.869020e-01    0.113098
source_name                    1.000000  2.196767e-03    0.997803
language_code                  1.000000  5.147065e-07    0.999999
region_codes                   1.000000  2.923275e-02    0.970767
company_codes_association      1.000000  5.147065e-07    0.999999
person_codes                   1.000000  8.579899e-02    0.914201
byline                         1.000000  3.056430e-02    0.969436
company_codes_relevance        1.000000  2.585252e-01    0.741475
source_code                    1.000000  2.105149e-03    0.997895
an                             1.000000  1.000000e+00    0.000000
word_count                     1.000000  7.242949e-03    0.992757
company_codes                  1.000000  3.621789e-01    0.637821
industry_codes                 1.000000  6.867213e-02    0.931328
title                          1.000000  7.835670e-01    0.216433
publication_datetime           1.000000  2.116277e-01    0.788372
publisher_name                 1.000000  7.339714e-04    0.999266
action                         1.000000  1.029413e-06    0.999999
document_type                  1.000000  5.147065e-07    0.999999
section                        0.448260  6.175186e-03    0.997231
dateline                       0.013851  1.921519e-01    0.997338

In [28]:
df['Non-duplicates'] = [set(code) for code in df['company_codes'].str.split(",")]

In [29]:
df['Non-duplicates'] = [list(code) for code in df['Non-duplicates']]

In [30]:
df['Non-duplicates'] = [','.join(code) for code in df['Non-duplicates']]

In [32]:
df['Non-duplicates'] = df['Non-duplicates'].str[1:]

In [33]:
df[['company_codes', 'Non-duplicates']]

company_codes  \
0          ,toamsi,toamsi,tosmsc,tosmsc,tshba,tshba,tshba,   
1                                      ,irevs,irevs,irevs,   
2                                                 ,euruno,   
3                   ,retrac,nyse,nyse,retrac,retrac,seexc,   
4                                   ,brstmy,brstmy,brstmy,   
...                                                    ...   
1942850                       ,egalcu,egalcu,egalcu,seexc,   
1942851  ,endoph,schplo,endoph,endoph,jnsspi,jnsspi,jon...   
1942852                                           ,amchso,   
1942853                     ,grnlft,kalamc,stry,stry,stry,   
1942854         ,zogeni,ardgm,cowhrp,zogeni,zogeni,endoph,   

                            Non-duplicates  
0                      tshba,toamsi,tosmsc  
1                                    irevs  
2                                   euruno  
3                        nyse,retrac,seexc  
4                                   brstmy  
...                                    ...  
1942850                       seexc,egalcu  
1942851  schplo,endoph,jnsspi,seexc,jonjon  
1942852                             amchso  
1942853                 grnlft,stry,kalamc  
1942854         zogeni,endoph,cowhrp,ardgm  

[1942855 rows x 2 columns]

To conclude, we found that the DNA data set includes 35 different variables with 1942855 entries. 4.5% of those entries were missing an actual value. Additionally, we tested seven variables for validity. The validation criteria depended uniquely on the variable. Overall, we found __ % of the entire data set to pass all of our validity checks. 

In [ ]:
### IAN's code from thursday morning 

In [ ]:
x = df[['company_codes', 'Non-duplicates']]

In [ ]:
my_list = []
for code in x['company_codes'].str.split(","):
    for code2 in code:
        my_list.append(code2)

In [ ]:
company_freq = pd.DataFrame()
company_code = np.array([])
company_count = np.array([])

In [ ]:
for company, count in Counter(my_list).items():
    company_code = np.append(company_code, company)
    company_count = np.append(company_count, count)


In [ ]:
company_freq['Code'] = company_code
company_freq['Count'] = company_count

In [ ]:
company_freq = company_freq.sort_values("Count", ascending = False).iloc[1:]


In [ ]:
company_freq.set_index("Code", inplace = True)


In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline
company_freq.Count.head(15).plot(kind = 'bar')
plt.ylabel("Frequency")
plt.title("Top 15 Most Mentioned Companines")


In [ ]:
#Made a csv of all companies with their frequency counts


In [ ]:
company_freq.to_csv("../data/working/companyfrequency.csv")


In [ ]:
company_freq.reset_index(inplace = True)
# Made a csv of all companies with their frequency counts and 
# corresponding company names(invalid companies will have NaN)


In [ ]:
name = np.array([])
val = pd.read_csv("../data/working/validcompaniesdictionary.csv", index_col = [0])

In [ ]:
val

In [ ]:
company_freq

In [ ]:
company_freq['Code'] = company_freq['Code'].str.upper()


In [ ]:
#Merging the valid company codes df with the company_freq df. This will match each company with their name, unless invalid
merged_df = pd.merge(val, company_freq, left_on = "Code", right_on = "Code", how = "outer")

In [ ]:
merged_df

In [ ]:
merged_df.to_csv("../data/working/companyfrequencyandname.csv")


In [ ]:
final = pd.read_csv("../data/working/companyfrequencyandname.csv", index_col = [0])
final.head(20)


In [ ]:
final[final.Code == 'PFIZ']
